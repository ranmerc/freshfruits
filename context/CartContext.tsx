import { useLocalStorage } from "usehooks-ts";
import {
  createContext,
  useContext,
  ReactNode,
  useReducer,
  Dispatch,
  useCallback,
} from "react";
import CartItem from "@/types/CartItem";

type ActionType =
  | {
      type: "ADD_TO_CART";
      payload: CartItem;
    }
  | {
      type: "REMOVE_FROM_CART";
      payload: {
        fruitId: number;
        selectedPackId: number;
      };
    }
  | {
      type: "CLEAR_CART";
    };

interface CartContextType {
  cartItems: CartItem[];
  cartDispatch: Dispatch<ActionType>;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export function useCartContext() {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }

  return context;
}

function cartReducer(state: CartItem[], action: ActionType) {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { fruitId, selectedPackId, quantity } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex !== -1) {
        const newState = [...state];
        newState[existingItemIndex].quantity += quantity;
        return newState;
      }

      return [...state, action.payload];
    }

    case "REMOVE_FROM_CART": {
      const { fruitId, selectedPackId } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex === -1) {
        return state;
      }

      const newState = [...state];
      newState[existingItemIndex].quantity -= 1;

      if (newState[existingItemIndex].quantity === 0) {
        newState.splice(existingItemIndex, 1);
      }

      return newState;
    }

    case "CLEAR_CART":
      return [];
    default:
      return state;
  }
}

export function CartProvider({ children }: { children: ReactNode }) {
  const [savedCartItems, setSavedCartItems] = useLocalStorage<CartItem[]>(
    "cart-items",
    []
  );

  const reducerLocalStorage = useCallback(
    (state: CartItem[], action: ActionType) => {
      const newState = cartReducer(state, action);

      setSavedCartItems(newState);

      return newState;
    },
    [setSavedCartItems]
  );

  const [state, dispatch] = useReducer(reducerLocalStorage, savedCartItems);

  return (
    <CartContext.Provider value={{ cartItems: state, cartDispatch: dispatch }}>
      {children}
    </CartContext.Provider>
  );
}
