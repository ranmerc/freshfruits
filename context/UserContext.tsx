import { createContext, useContext, ReactNode } from "react";
import { useLocalStorage } from "usehooks-ts";

export interface UserContextType {
  username: string;
  setUserName: (user: string) => void;
}

export const UserContext = createContext<UserContextType | undefined>(
  undefined
);

export function useUserContext() {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error("useUserContext must be used within a UserProvider");
  }

  return context;
}

export function UserProvider({ children }: { children: ReactNode }) {
  const [username, setUserName] = useLocalStorage<string>("username", "");

  return (
    <UserContext.Provider value={{ username, setUserName }}>
      {children}
    </UserContext.Provider>
  );
}
