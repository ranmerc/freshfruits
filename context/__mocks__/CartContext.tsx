export const CartContext = jest.fn();

export const useCartContext = jest.fn();

export const CartProvider = jest.fn();
