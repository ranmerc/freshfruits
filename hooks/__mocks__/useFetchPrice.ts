const useFetchPrice = jest.fn(() => {});

export default useFetchPrice;
