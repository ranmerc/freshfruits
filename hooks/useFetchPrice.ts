import PriceData from "@/types/PriceData";
import axios from "axios";
import { useQuery } from "@tanstack/react-query";

const fetcher = async (id: number) => {
  const response = await axios.get<PriceData>(`/api/pricing?id=${id}`);
  return response.data;
};

export default function useFetchPrice(id: number) {
  return useQuery<PriceData>({
    queryKey: ["fruit", id],
    queryFn: () => fetcher(id),
  });
}
