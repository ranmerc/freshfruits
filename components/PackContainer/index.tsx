import useFetchPrice from "@/hooks/useFetchPrice";
import {
  Alert,
  Button,
  CircularProgress,
  Stack,
  Typography,
} from "@mui/material";
import PricePack from "../PricePack";
import { ChangeEvent, useState } from "react";
import AddShoppingCartIcon from "@mui/icons-material/AddShoppingCart";

export default function PackContainer({ id }: { id: number }) {
  const { data, isLoading, isError, isSuccess } = useFetchPrice(id);
  const [selectedPack, setSelectedPack] = useState<number | null>(null);

  const handleChange: (
    event: ChangeEvent<HTMLInputElement>,
    checked: boolean
  ) => void = (e) => {
    setSelectedPack(Number(e.target.value));
  };

  if (isLoading) {
    return (
      <Stack alignItems={"center"}>
        <CircularProgress color="secondary" aria-label="Loading prices" />
      </Stack>
    );
  }

  if (isError) {
    return (
      <>
        <Alert variant="filled" severity="error">
          Unable to load prices!
        </Alert>
      </>
    );
  }

  if (isSuccess && data) {
    return (
      <Stack
        component={"form"}
        onSubmit={(e) => {
          e.preventDefault();
          if (selectedPack !== null) {
            console.log(data.packs[selectedPack]);
          }
        }}
        rowGap={{ xs: 1, md: 2 }}
      >
        <Typography variant="body1" component={"h3"} fontWeight={"bold"}>
          Select Pack
        </Typography>
        <Stack flexDirection={"row"}>
          {data.packs.map((pack, i) => {
            return (
              <PricePack
                info={pack}
                selected={selectedPack}
                onChange={handleChange}
                packId={i}
                key={i}
              />
            );
          })}
        </Stack>
        <Button
          variant="contained"
          type="submit"
          sx={{
            py: 1.5,
            fontSize: "1rem",
            width: { sx: "100%", md: "fit-content" },
          }}
          startIcon={<AddShoppingCartIcon />}
          disabled={selectedPack === null}
        >
          Add to Cart
        </Button>
      </Stack>
    );
  }
}
