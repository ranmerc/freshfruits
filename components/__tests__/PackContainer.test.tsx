import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import PackContainer from "../PackContainer";
import useFetchPrice from "@/hooks/useFetchPrice";
import { useCartContext } from "@/context/CartContext";
import renderer from "react-test-renderer";

// telling jest to load the mock implementation from __mocks__
jest.mock("@/hooks/useFetchPrice");
jest.mock("@/context/CartContext");

// satisfying ts error no mockImplementation
const mocked = useFetchPrice as jest.Mock;
const mockedUseCartContext = useCartContext as jest.Mock;
mockedUseCartContext.mockReturnValue({
  addItemToCart: jest.fn(),
});

describe("Pack container", () => {
  it("Renders loading view", () => {
    mocked.mockImplementation(() => ({
      isLoading: true,
      isError: false,
      data: null,
      isSuccess: false,
    }));

    render(<PackContainer id={1} />);

    expect(
      screen.getByRole("progressbar", { name: "Loading prices" })
    ).toBeInTheDocument();

    const tree = renderer.create(<PackContainer id={1} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders error view", () => {
    mocked.mockImplementation(() => ({
      isLoading: false,
      isError: true,
      data: null,
      isSuccess: false,
    }));
    render(<PackContainer id={1} />);

    expect(screen.getByRole("alert")).toBeInTheDocument();
    expect(screen.getByText("Unable to load prices!")).toBeInTheDocument();

    const tree = renderer.create(<PackContainer id={1} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders heading", () => {
    mocked.mockImplementation(() => ({
      isLoading: false,
      isError: false,
      data: {
        id: 1,
        packs: [
          {
            type: "count",
            quantity: 1,
            price: 70,
            discountPrice: 70,
            inStock: true,
          },
          {
            type: "count",
            quantity: 6,
            price: 400,
            discountPrice: 300,
            inStock: false,
          },
        ],
      },
      isSuccess: true,
    }));

    render(<PackContainer id={1} />);
    expect(screen.getByRole("heading", { level: 3 })).toBeInTheDocument();
  });

  it("Renders price packs", () => {
    mocked.mockImplementation(() => ({
      isLoading: false,
      isError: false,
      data: {
        id: 1,
        packs: [
          {
            type: "count",
            quantity: 1,
            price: 70,
            discountPrice: 70,
            inStock: true,
          },
          {
            type: "count",
            quantity: 6,
            price: 400,
            discountPrice: 300,
            inStock: false,
          },
        ],
      },
      isSuccess: true,
    }));

    render(<PackContainer id={1} />);

    const radios = screen.getAllByRole("radio");
    expect(radios).toHaveLength(2);

    const tree = renderer.create(<PackContainer id={1} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
