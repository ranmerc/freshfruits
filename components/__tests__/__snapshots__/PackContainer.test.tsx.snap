// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pack container Renders error view 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-filledError MuiAlert-filled css-1c3trtb-MuiPaper-root-MuiAlert-root"
  role="alert"
>
  <div
    className="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
      data-testid="ErrorOutlineIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
  </div>
  <div
    className="MuiAlert-message css-1pxa9xg-MuiAlert-message"
  >
    Unable to load prices!
  </div>
</div>
`;

exports[`Pack container Renders loading view 1`] = `
<div
  className="MuiStack-root css-163ljxj-MuiStack-root"
>
  <span
    aria-label="Loading prices"
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSecondary css-1szo49z-MuiCircularProgress-root"
    role="progressbar"
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={{}}
      />
    </svg>
  </span>
</div>
`;

exports[`Pack container Renders price packs 1`] = `
<form
  className="MuiStack-root css-1cgqb3-MuiStack-root"
  onSubmit={[Function]}
>
  <h3
    className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
  >
    Select Pack
  </h3>
  <div
    className="MuiStack-root css-ku2emo-MuiStack-root"
  >
    <span
      className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <input
        aria-label="1 for 70"
        checked={false}
        className="PrivateSwitchBase-input css-1m9pwf3"
        disabled={false}
        name="price-packs"
        onChange={[Function]}
        required={false}
        type="radio"
        value={0}
      />
      <div
        className="MuiBox-root css-0"
      >
        <div
          className="MuiBox-root css-4f8iwy"
        >
          <span
            className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
          >
            Fixed Price
          </span>
        </div>
        <div
          className="MuiStack-root css-ah0cgr-MuiStack-root"
        >
          <div
            className="MuiStack-root css-1qdfthb-MuiStack-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              ₹
              70
            </p>
          </div>
          <span
            className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
          >
            1
             
            Nos
          </span>
        </div>
      </div>
    </span>
    <span
      aria-disabled={true}
      className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary Mui-disabled PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-disabled MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
    >
      <input
        aria-label="6 for 300"
        checked={false}
        className="PrivateSwitchBase-input css-1m9pwf3"
        disabled={true}
        name="price-packs"
        onChange={[Function]}
        required={false}
        type="radio"
        value={1}
      />
      <div
        className="MuiBox-root css-0"
      >
        <div
          className="MuiBox-root css-12xkctz"
        >
          <span
            className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
          >
            Out of Stock
          </span>
        </div>
        <div
          className="MuiStack-root css-ah0cgr-MuiStack-root"
        >
          <div
            className="MuiStack-root css-1qdfthb-MuiStack-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
            >
              ₹
              300
            </p>
            <p
              className="MuiTypography-root MuiTypography-body2 css-qc7gyi-MuiTypography-root"
            >
              ₹
              400
            </p>
          </div>
          <span
            className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
          >
            6
             
            Nos
          </span>
        </div>
      </div>
    </span>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-10ihbrb-MuiButtonBase-root-MuiButton-root"
    disabled={true}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="submit"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddShoppingCartIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
        />
      </svg>
    </span>
    Add to Cart
  </button>
</form>
`;
